### ISSUE TITLE
Formatting issue in usage section with S4 methods

### ISSUE THREAD

**Issue by andrewmarx** [r-lib/roxygen2#896]
I'm having issues where sometimes Roxygen is breaking method parameters across lines for my S4 methods in the usage section of my Rd files. I can't state for sure if this is specific to S4 methods or not. Rarely, I'll even see extra blank lines or indentation before \S4method{}, but this behavior is very inconsistent. An example of the former issue at least is seen here: 

https://github.com/andrewmarx/samc/blob/fd5a02615a624c361cbe8421153169cdb5db3652/man/distribution.Rd#L13-L32

The code this was generated from is available here:

https://github.com/andrewmarx/samc/blob/fd5a02615a624c361cbe8421153169cdb5db3652/R/distribution.R

And here is a screenshot of the issue (output from pkgdown):

![distribution](https://user-images.githubusercontent.com/13956617/63361038-6cf61a00-c33d-11e9-9193-89f33239f61e.png)

If I had to hazard a guess, it would be that Roxygen is trying to respect some preconceived line width when writing the \S4method{} lines to the Rd files, but is not accounting for how the different components of the \S4method{} lines will actually be displayed in the final documentation.

**Comment by hadley** [comment]
Yes, that's exactly the problem. I think this is most likely to be fixed via #820 where we (optionally) switch to putting one argument per line. Otherwise resolving this problem requires a significant upgrade to the data model underlying usage wrapping, and that's unlikely to be worth it.

**Comment by andrewmarx** [comment]
Yes, I traced everything through to the wrapString() C++ function and can see how it would be a hassle and inelegant to code for corner cases in it. Perhaps a separate wrapUsage() function to be called by topic_add_usage() instead of wrapString(), but I digress.

I reopened this topic because I isolated the cause of the other issue I mentioned (errant newline and indent). In the wrapString() function, if the first piece of text is longer than the line width, the output gets a newline and indent prepended.

```
wrapString("\\S4method{dispersal}{samc,missing,missing,numeric,numeric}(samc, dest, time)", width = 30)

[1] "\n  \\S4method{dispersal}{samc,missing,missing,numeric,numeric}(samc,\n  dest, time)"
```

I suspect this unintended behavior (and very rarely encountered). One potential fix involves an if statement and starting the loop at `i = 1`

```
std::string wrapString(std::string string, int width = 80, int indent = 2) {
  std::vector<std::string> pieces = splitByWhitespace(string);
  int n = pieces.size();
  int cur_width = 0;
  
  std::string out;
  
  if (n > 0) {
    cur_width = pieces[0].size();
    out = pieces[0];
  }
  
  for (int i = 1; i < n; ++i) {
    int piece_width = pieces[i].size();
    
    if (piece_width + cur_width < width) {
      cur_width += piece_width;
      if (i != 0) {
        out += " ";
        cur_width++;
      }
    } else {
      cur_width = piece_width + indent;
      out += "\n" + std::string(indent, ' ') ;
    }
    out += pieces[i];
  }
  
  return out;
}
```

I don't have a comprehensive understanding of the package or the situations where wrapString() is used, and I can't test it comprehensively right now to make sure it doesn't break anything, so I'm a little hesitant to implement it myself as a pull request

**Comment by andrewmarx** [comment]
Actually, I just realized wrap_string() doesn't appear to be used outside of topic_add_usage(), so it could potentially be modified directly. Assuming you don't want to risk breaking compatibility for anyone that else that might be using it, you could alternatively add something like this:

```
// [[Rcpp::export]]
std::vector<std::string> splitByBrace(std::string string) {
  std::vector<std::string> out;
  
  std::string acc = "";
  char in_string = '\0';
  int in_escape = 0;
  
  std::string::const_iterator cur = string.begin(), end = string.end();
  
  while(cur != end) {
    if (in_string != '\0') {
      acc += *cur;
      
      if (in_escape) {
        in_escape--;
      } else if (*cur == '\\' && cur + 1 != end && *(cur + 1) == '\\') {
        in_escape = 2;
      } else if (*cur == in_string) {
        // String terminates
        in_string = '\0';
      }
      
    } else if (*cur == '{' || *cur == '}') {
      out.push_back(acc);
      acc = "";
    } else if (*cur == '"' || *cur == '\'') {
      in_string = *cur;
      acc += *cur;
    } else {
      acc += *cur;
    }
    
    cur++;
  }
  
  out.push_back(acc);
  
  return out;
}
```

and

```
// [[Rcpp::export]]
std::string wrapUsage(std::string string, int width = 80, int indent = 2) {
  std::vector<std::string> pieces = splitByBrace(string);
  int n = pieces.size();
  int cur_width = 0;
  
  std::string out;

  if (n > 0) {
    if (pieces[0] == "\\method" || pieces[0] == "\\S3method" || pieces[0] == "\\S4method") {
      out = pieces[0] + 
        "{" + pieces[1] + "}{" + pieces[3] + "}" + //pieces[2] is an empty string that goes between the }{, so we just ignore it
        wrapString(pieces[4], width - pieces[1].size(), indent);
    } else {
      // Just a treat things like they were before
      out = wrapString(string, width, indent);
    }
  }

  return out;
}
```

It's just a quick and dirty proof of concept that appears to work with the following caveats:
- I've only tested wrapUsage() directly, not as part of the whole package
- It assumes \method, \S3method, and \S4method always have the exact same structure (with two sets of braces). They may need to be in separated into separate if else blocks
- I don't know all the different things that can go into the \usage{} section of an Rd file, so it may have to be adjusted to account for other things.

**Comment by hadley** [comment]
That seems promising — thanks!

**Comment by hadley** [comment]
Now that I've implemented #820, which I'm reasonably confident that most people will prefer (since it makes it much easier to skim the arguments), I no longer think it's worth fixing this.
